### PR TITLE
feat: Add benchmarks for exprtk evaluator.

### PR DESCRIPTION
**Pull request by sheny1xuan**
+ [feat: exprtk evaluator support RBAC with domain](https://github.com/casbin/casbin-cpp/commit/f7f7e3b83e79627a4653a481deea28eb89cbc389)
+ [feat: exprtk evaluator support rbac with pattern](https://github.com/casbin/casbin-cpp/commit/f0053dc88b83d6d726be683ab0214bc84041c968)
+ [feat: add benchmark for exprtk evaluator](https://github.com/casbin/casbin-cpp/commit/711c3b20a894661199c553ee3e19be0d0ac96e4a)
+ [perf: only compile once in Eval when eval string don't change.](https://github.com/casbin/casbin-cpp/commit/82bfac461b4e9c11454923a35e931c5092507cc9)
+ [fix: repair regex function and add function judge whether's null](https://github.com/casbin/casbin-cpp/commit/1bac8f190d7057a664f542ab1cfe472a5e24c9ab)
+ [fix: clean all symbol table to avoid haning pointer in exprtk.](https://github.com/casbin/casbin-cpp/commit/70b4b7749205fd356199fbd3bcd1c4c58cb7d386)

+ Exprt performance is much better than duktape, and we also have many  performace enhancement work to do with exprtk evaluator. And we need to wait exprtk new [feature](https://github.com/casbin/casbin-cpp/pull/185#issuecomment-1055326799), and we can achieve zero copy for enforce data in this way.
image
+ I think maybe we should remove duktape evaluator, it's inefficient. @EmperorYP7

### CODE REVIEW

**Comment by hsluoyz** [comment]
@EmperorYP7

File: casbin/model/evaluator.cpp [hunk 1]
```diff
@@ -104,21 +115,24 @@ namespace casbin {
         for (auto& var: var_list) {
             printf(" %s: %s\n" , var.c_str(), symbol_table.get_stringvar(var)->ref().c_str());
         }
+        printf("Current error: %s\n", parser.error().c_str());
+        // printf("Current exprsio string: %s\n", parser.current_token);
+        printf("Current value: %d\n", bool(this->expression));
```
Comment (EmperorYP7) [inline casbin/model/evaluator.cpp:120]: We might look into a new logging system for the project

File: include/casbin/model/exprtk_config.h [hunk 2]
```diff
@@ -98,8 +100,70 @@ namespace casbin {
...
+    public:
+        ExprtkOtherFunction(const std::string& idenfier, casbin::MatchingFunc func)
+        : exprtk::igeneric_function<numerical_type>(idenfier), func_(func)
+        {}
+
+        ExprtkOtherFunction()
+        : exprtk::igeneric_function<numerical_type>("ss")
+        {}
+
+        inline numerical_type operator()(parameter_list_t parameters) {
```
Comment (EmperorYP7) [inline include/casbin/model/exprtk_config.h:126]: Excellent work!

File: include/casbin/model/exprtk_config.h [hunk 3]
```diff
@@ -108,9 +172,28 @@ namespace casbin {
                 if (type == ExprtkFunctionType::Gfunction) {
                     std::string idenfier(narg, 'S');
                     return std::make_shared<ExprtkGFunction>(idenfier, rm);
-                } else {
-                    return nullptr;
+                } else if (type != ExprtkFunctionType::Unknown) {
+                    std::string idenfier(narg, 'S');
+                    auto ret = std::make_shared<ExprtkOtherFunction>();
+                    if (type == ExprtkFunctionType::KeyMatch) {
```
Comment (EmperorYP7) [inline include/casbin/model/exprtk_config.h:178]:
```suggestion
                    switch (type) {
```

**Review by EmperorYP7 — APPROVED** [review]
Thoroughly satisfied with the results. Well done @sheny1xuan!

File: include/casbin/model/exprtk_config.h (hunk 3 above)
Comment (sheny1xuan) [reply include/casbin/model/exprtk_config.h]: I have changed. But this commit makes these conflict, sorry about that.
image

**Comment by sheny1xuan** [comment]
I think we can remove duktape evaluator now, exprtk evaluator is able to complete the task. @EmperorYP7

**Comment by sheny1xuan** [comment]
Could you merge this pr, so I can continue to remove the old evaluator based on this. @hsluoyz

**Comment by hsluoyz** [comment]
@sheny1xuan plz resolve conflicts first:

image

**Comment by hsluoyz** [comment]
@sheny1xuan you don't need to force-write into this repo. You only need to update your own fork code. Pull the latest code, resolve the conflicts locally, push into your fork. Then it's OK.

image

**Comment by sheny1xuan** [comment]
> @sheny1xuan you don't need to force-write into this repo. You only need to update your own fork code. Pull the latest code, resolve the conflicts locally, push into your fork. Then it's OK.
> 
> image

Thanks, have fixed it.

**Pull request merged**